src/lib/reasoners/uf.ml

```diff
@@ -999,100 +999,206 @@ let assign_next env =
-        with Not_found -> assert false
-      in
-      let cls =
-        try List.rev_map (fun s -> s, ME.find s env.make) cls
-        with Not_found -> assert false
-      in
-      let e = X.choose_adequate_model t rep cls in
-      e, ME.add t e mrepr
+    let (rep, hs), mrepr =
+      match unbounded with
+      | Some string_repr ->
+        (rep, Hstring.make string_repr), mrepr
+      | None ->
+        let cls =
+          try SE.elements (MapX.find rep env.classes)
+          with Not_found -> assert false
+        in
+        let cls =
+          try List.rev_map (fun s -> s, ME.find s env.make) cls
+          with Not_found -> assert false
+        in
+        let rep, string_repr = X.choose_adequate_model t rep cls in
+        let hs = Hstring.make string_repr in
+        (rep, hs), ME.add t (rep, hs) mrepr
+    in
+    assert (is_a_good_model_value rep && is_const_term rep);
+    (rep, hs), mrepr
 
 let is_optimization_op = function
   | Sy.Op Sy.Optimize _ -> true
   | _ -> false
 
-let save_cache () =
-  LX.save_cache ()
+(* Caches used during the computation of the model.
+   All these caches are flushed before generating a new model. *)
+module Cache = struct
+  (* Stores all the get accesses to arrays. *)
+  let arrays_cache = Hashtbl.create 17
 
-let reinit_cache () =
-  LX.reinit_cache ()
+  (* Stores all the accesses to records. *)
+  let records_cache = Hashtbl.create 17
 
-let compute_concrete_model_of_val
-    env t ((fprofs, cprofs, carrays, mrepr) as acc) unbounded =
-  let { E.f; xs; ty; _ } = E.term_view t in
+  (* Stores all the abstract values generated. This cache is necessary
+     to ensure we don't generate twice an abstract value for a given symbol. *)
+  let abstracts_cache = Hashtbl.create 17
+
+  let store_array_get t i v =
+    match Hashtbl.find_opt arrays_cache t with
+    | Some values ->
+      Hashtbl.replace values i v
+    | None ->
+      let values = Hashtbl.create 17 in
+      Hashtbl.add values i v;
+      Hashtbl.add arrays_cache t values
+
+  let store_record_sig id ty =
+    match Hashtbl.find_opt records_cache (id, ty) with
+    | Some _ -> ()
+    | None ->
+      let values = Hashtbl.create 17 in
+      Hashtbl.add records_cache (id, ty) values
+
+  let store_record_access id ty field v =
+    match Hashtbl.find_opt records_cache (id, ty) with
+    | Some values ->
+      Hashtbl.replace values field v
+    | None ->
+      assert false
+
+  let get_abstract_for env t =
+    let r, _ = find env t in
+    match Hashtbl.find_opt abstracts_cache r with
+    | Some abstract -> abstract
+    | None ->
+      let abstract = Id.Namespace.Abstract.fresh () |> Hstring.make in
+      Hashtbl.add abstracts_cache r abstract;
+      abstract
+
+  let clear () =
+    Hashtbl.clear arrays_cache;
+    Hashtbl.clear records_cache;
+    Hashtbl.clear abstracts_cache
+end
+
+let is_forbidden_symbol f ty =
   (* Keep record constructors because models.ml expects them to be there *)
```

Comment:
The corresponding code that expects the record constructors to be there in `models.ml` seems to have been removed, was it intentional?

Reply:
Uhm I'm not sure. The function `is_forbidden_symbol` is used to filter the symbols on which we will iterate during the model generation. I think the comment means we have to keep `Access` symbol, even if this symbol is interpreted by the theory of records, otherwise we cannot construct the model value for this theory.

Reply:
I wrote this comment. It refers to the fact that in we (used to) need to preserve the record constructors for `check_records` et al in `models.ml`, but those functions have been removed. This means that either some stuff will no longer be working, or the new code in `uf.ml` is meant to replace those functions and we don't need to keep the record constructors here anymore (probably the latter).